Refactored WebAuthn with Windows Hello support

nikosdion opened the pull request:
### Summary of Changes

* Refactored as a native Joomla 4 plugin, using a service provider and implementing SubscriberInterface
* Removed the ugly `Joomla` helper class, replacing it with native code
* Now using the (previously undocumented) WebAuthn library's `WebAuth\Server` object. This adds Windows Hello support **without** having to update to a new major version of the third party WebAuthn library
* We no longer pass URLs as data arguments; we can figure them out using `Joomla.getOptions`
* Fixed the breakage introduced in https://github.com/joomla/joomla-cms/pull/37464

The PR replaces #37673 and #37675

### Testing Instructions

Please remember to run `npm ci` after applying the PR — the JavaScript has changed.

Please remember to use HTTPS with a certificate trusted by your computer; WebAuthn doesn't work on plain HTTP.

Please use a relatively recent (2019 onwards) build of Chrome, Edge, Firefox etc.

Go to your user profile in the backend of the site.

Click on the ‘W3C Web Authentication (WebAuthn)’ tab.

On a Windows computer **without** any hardware authenticator attached click on Add New Authenticator.

### Actual result BEFORE applying this Pull Request

The browser asks you to plug in an authenticator.

### Expected result AFTER applying this Pull Request

You can enter your PIN / show your face / use a fingerprint scanner to register Windows Hello as an authenticator.

### Further testing

Delete the authenticator and try adding it again in the user profile page in the frontend of the site. It should still work.

Make sure that in the frontend you can delete an authenticator you added in the backend.

Make sure that in the backend you can delete an authenticator you added in the frontend.

Please make sure you can add more than one authenticators. **IMPORTANT!** You cannot add the same authenticator twice (in the past you could; it was a bug that went unnoticed). You can only test this if you have more than one authenticators, e.g. Windows Hello, a FIDO or FIDO2 hardware authenticator, an Android phone and so on.

Please make sure that you can edit the name of the authenticator. This was broken in #37464 and was still broken at the time I made this PR.

Please make sure you can log into the front- and backend of the site.

Please test on as many platforms as you have: Android (works on Android 9 and later if you have a fingerprint scanner but only on Chrome as far as I know), iOS/iPadOS (both TouchID and FaceID), macOS (TouchID, if you have a MacBook Air/Pro or an iMac/Mac Studio with Apple Silicon and the Apple keyboard with a TouchID sensor) as well as various FIDO and FIDO2 authenticators. I have tested all of these and Linux ~EXCEPT for Android due to lack of hardware running Android (my Android phone's battery bloated, I had to decommission it before it spontaneously turned into an incendiary grenade)~. Edit: I now have a Samsung Galaxy A21 device for testing with face recognition and fingerprint scanner which I used to confirm it works with Android.

### Translation strings

The following language strings were added:

* `PLG_SYSTEM_WEBAUTHN_ERR_XHR_INITCREATE`
* `PLG_SYSTEM_WEBAUTHN_FIELD_ATTESTATION_SUPPORT_DESC`
* `PLG_SYSTEM_WEBAUTHN_FIELD_ATTESTATION_SUPPORT_LABEL`
* `PLG_SYSTEM_WEBAUTHN_LBL_DEFAULT_AUTHENTICATOR`

The following language strings were changed:

* `PLG_SYSTEM_WEBAUTHN_LBL_DEFAULT_AUTHENTICATOR_LABEL`

### Documentation Changes Required

As of Joomla! __DEPLOY_VERSION__ the WebAuthn plugin has attestation enabled by default. This means that only authenticators with publicly verifiable cryptographic signatures can be registered with WebAuthn starting with this version of Joomla.

The publicly verifiable certification authorities for authenticators are retrieved from the FIDO Alliance site, namely the URL `https://mds.fidoalliance.org/`.

This default setting **will** prevent some cheaper authenticators which are not FIDO-certified from being used with WebAuthn. Moreover, some sites may be unable to download and/or cache the root certificates from FIDO Alliance, or it might take so long that the plugin aborts the operation to prevent your site from timing out. If you encounter any problems with registering authenticators with WebAuthn please edit the plugin settings and disable the Attestation Support option.

The Attestation Support feature requires the following prerequisites to work:

1. Your site must be able to access directly the URL `https://mds.fidoalliance.org/`.
2. Your cache folder (`administrator/cache`) must be writeable by PHP.
3. The system temporary directory must be writeable by PHP.
4. The OpenSSL extension must be installed and enabled on your site — this is a requirement for WebAuthn as a whole.

If these prerequisites are not met the WebAuthn plugin will proceed _without_ verifying the cryptographic signatures of the authenticators against the publicly verifiable certification authorities published by the FIDO Alliance. This is still secure — in fact far more secure than using a password and Two Factor Authentication. The only downside is that you may experience a short delay, up to 5 seconds, once a month when the plugin attempts to download the root certification authority information from the FIDO Alliance.

If your site meets all of the prerequisites except the first one you may download the information from `https://mds.fidoalliance.org/` and place them in the file `administrator/cache/fido.jwt`. In this case the WebAuthn plugin can operate with attestation support. This is very useful if your site is behind a firewall or disconnected from the Internet (e.g. on a high security intranet handling sensitive material). You need to remember to update this file once every month to avoid any problems.

Enabling the Attestation Support feature also allows Joomla to identify the maker and model of most FIDO2 certified authenticators. If you register an authenticator _after_ enabling this option you will see an icon of the maker's logo next to the Authenticator Name when viewing the list of authenticators. Furthermore, registering a new authenticator will have a more user-friendly default name, e.g. “Yubikey 5Ci added on 28 April 2022, 18:00” instead of “Authenticator added on 28 April 2022, 18:00”. 

If you disable the Attestation Support option the logo and the authenticator type will be hidden.

Finally, do note that authenticators added with previous versions of Joomla or while the Attestation Support feature is disabled or while the Attestation Support feature is enabled but its prerequisites not met will _always_ be displayed as “Generic Authenticator” as the necessary information to determine the make and model of the authenticator will have not been relayed to Joomla when you registered your authenticator.

Commit:
Refactored WebAuthn plugin

Commit:
Fix the WebAuthn management page which was broken in https://github.com/joomla/joomla-cms/pull/37464

Commit:
Fix wrong `@since` doc tag

Commit:
Fix docblock typo

Commit:
Fix docblock typo

Commit:
Fix docblock typo

Commit:
Fix docblock typo

Commit:
Fix docblock typo

Commit:
Fix broken management interface

Commit:
Make unnecessarily static method back into non-static

Commit:
Replace static helper with injected object

Commit:
Come on, commit the ENTIRE file!

Commit:
Use the user factory

Commit:
Fix error when going through the user factory

Commit:
Fix: cannot add WebAuthn authenticator right back after deleting it

Commit:
Remove useless switch branch

Commit:
Remove useless exception

Commit:
Display make and model of the authenticator, if possible

Commit:
Add missing JWT signature algorithms

Commit:
Fix copyright date

Commit:
Fix for PHP 8 using FIDO keys and Android phones

Commit:
Reactivate the tooltips after adding an authenticator

Commit:
Option to disable the attestation support

Commit:
The Windows Hello icon was invisible on white background

Commit:
Attempt to fix Appveyor not having Sodium in the Windows build

Commit:
Work around third party library bug...

richard67 committed:
Merge branch '4.2-dev' into feature/webauthn-refactor-2

Commit:
Merge branch '4.2-dev' into feature/webauthn-refactor-2

Commit:
Create events in a forwards-compatible manner

Commit:
Concrete events

Commit:
Merge branch '4.2-dev' into feature/webauthn-refactor-2

richard67 committed:
Merge branch '4.2-dev' into feature/webauthn-refactor-2

brianteeman commented:
Trying to test this and I keep getting the error message "Cannot get the authenticator registration information from your site."

This could be because its a local server with a self signed cert in which case I will just move along but thought I should check with you first

nikosdion commented:
@brianteeman Windows, Android or something else? And which browser? It's been a while since I wrote the code, it'll be easier for me to test myself with a similar environment to the one you are using :)

brianteeman commented:
Testing on windows 11 with chrome.

The server is using [laragon ](https://laragon.org/)(which seems very fast) and the crt from latagon

php 8.1.6

nikosdion commented:
Are you accessing the local server as `localhost` or using a (locally resolved to `127.0.0.1`) domain name? In the former case I expect it to fail.

Commit:
Merge branch '4.2-dev' into feature/webauthn-refactor-2

Commit:
Fix event woes

nikosdion commented:
@brianteeman Fixed. Try again now, please.

brianteeman commented:
That problem is fixed. 

Next problem is

![image](https://user-images.githubusercontent.com/1296369/173238442-5b43660c-13f1-47ba-8916-f83949e3fe8f.png)

I kind of expected the error message (for reasons stated above) but not the deprecated messages

brianteeman commented on plugins/system/webauthn/webauthn.xml:
```diff
@@ -11,12 +11,31 @@
+		<fields name="params">
+			<fieldset name="basic">
+				<field
+						name="attestationSupport"
```
```suggestion
					name="attestationSupport"
```

brianteeman commented on plugins/system/webauthn/webauthn.xml:
```diff
@@ -11,12 +11,31 @@
+			<fieldset name="basic">
+				<field
+						name="attestationSupport"
+						type="radio"
```
```suggestion
					type="radio"
```

nikosdion commented:
Can't do anything about the deprecated warnings. As you can see, they come from third party libraries and I was explicitly told that Joomla's sever promise extends to third party libraries included via Composer even if they are only consumed internally.

The invalid certificate chain error happens because you do not have the PHP OpenSSL extension enabled and you do not have the openssl command line programme on your server either. Again, nothing I can do about.

nikosdion commented:
@brianteeman Also, FWIW, both the deprecated notice and the invalid certificate message will happen with the WebAuthn plugin as included in Joomla 4.0. They were NOT introduced with this PR. They come from third party code which as I already explained I'm not allowed to update...

nikosdion commented:
@brianteeman I also received an email about two suggestions you made... but they are nowhere to be found on GitHub?!

brianteeman commented on plugins/system/webauthn/webauthn.xml:
```diff
@@ -11,12 +11,31 @@
+		<fields name="params">
+			<fieldset name="basic">
+				<field
+						name="attestationSupport"
+						type="radio"
+						layout="joomla.form.field.radio.switcher"
+						label="PLG_SYSTEM_WEBAUTHN_FIELD_ATTESTATION_SUPPORT_LABEL"
+						description="PLG_SYSTEM_WEBAUTHN_FIELD_ATTESTATION_SUPPORT_DESC"
+						default="1"
+						filter="integer"
+				>
```
```suggestion
					name="attestationSupport"
					type="radio"
					label="PLG_SYSTEM_WEBAUTHN_FIELD_ATTESTATION_SUPPORT_LABEL"
					description="PLG_SYSTEM_WEBAUTHN_FIELD_ATTESTATION_SUPPORT_DESC"
					layout="joomla.form.field.radio.switcher"
					default="1"
					filter="integer"
					>
```

brianteeman commented on administrator/language/en-GB/plg_system_webauthn.ini:
```diff
@@ -18,17 +18,21 @@ PLG_SYSTEM_WEBAUTHN_ERR_CREDENTIAL_ID_ALREADY_IN_USE="Cannot save credentials. T
 PLG_SYSTEM_WEBAUTHN_ERR_USER_REMOVED="The user for this authenticator seems to no longer exist on this site."
+PLG_SYSTEM_WEBAUTHN_ERR_XHR_INITCREATE="Cannot get the authenticator registration information from your site."
+PLG_SYSTEM_WEBAUTHN_FIELD_ATTESTATION_SUPPORT_DESC="Only allow authenticators with verifiable cryptographic signatures to be used for WebAuthn logins. Strongly recommended for high security environments. Requires your site to be able to access <code>https://mds.fidoalliance.org/</code> directly, a writeable cache directory, the system temporary directory being writeable by PHP, and the OpenSSL ex ...
+PLG_SYSTEM_WEBAUTHN_FIELD_ATTESTATION_SUPPORT_LABEL="Attestation support"
```
```suggestion
PLG_SYSTEM_WEBAUTHN_FIELD_ATTESTATION_SUPPORT_LABEL="Attestation Support"
```

following the styleguide

brianteeman commented:
> Can't do anything about the deprecated warnings. As you can see, they come from third party libraries and I was explicitly told that Joomla's sever promise extends to third party libraries included via Composer even if they are only consumed internally.

ok

> The invalid certificate chain error happens because you do not have the PHP OpenSSL extension enabled and you do not have the openssl command line programme on your server either. Again, nothing I can do about.

What I expect - was just reporting. Disabling atestation resolved that

brianteeman commented:
> @brianteeman I also received an email about two suggestions you made... but they are nowhere to be found on GitHub?!

sorry - deleted them and replaced with a new suggestion

Commit:
Update plugins/system/webauthn/webauthn.xml

Co-authored-by: Brian Teeman <brian@teeman.net>

Commit:
Update administrator/language/en-GB/plg_system_webauthn.ini

Co-authored-by: Brian Teeman <brian@teeman.net>

richard67 commented:
@nikosdion We have 2 conflicting files now after the last upmerge from 4.1-dev to 4.2-dev, both caused by PR #37676 . Do you think you can fix them? Let me know if you need some help with that (but I'm sure you won't need some).

Commit:
Merge branch '4.2-dev' into feature/webauthn-refactor-2

Kostelano commented:
When editing an authenticator:

![Screenshot_1](https://user-images.githubusercontent.com/8440661/173432451-cfc5a8ab-8e89-4065-81d8-3cdc6e6f6a94.png)

Kostelano commented:
And, returning to the issue of confirming the deletion (as it was eventually implemented in the MFA), we delete the authenticator without confirmation. I would still prefer to have a second "intentional" click before deleting.

roland-d committed:
Merge branch '4.2-dev' into feature/webauthn-refactor-2

Commit:
Merge branch '4.2-dev' into feature/webauthn-refactor-2

Commit:
Improve the layout for editing an authenticator

It now follows the Bootstrap 5 form aesthetic. Moreover,
there are gaps between the text input and the Save and
Cancel buttons.

Commit:
Confirm deletion of authenticators

nikosdion commented:
@Kostelano Both points addressed :) If you're happy with it I think it's ready to merge.

richard67 commented:
I think it needs again 2 human tests.

nikosdion commented:
@richard67 Maybe @Kostelano and @brianteeman would be able to provide their successful tests again. The bulk of the code which actually introduces the new feature is unchanged the past couple of months, really.

Kostelano commented:
I have tested this item :white_check_mark: successfully on c03b191350d4003b9ca3ba05a0c4c423c6d52a35<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37910">issues.joomla.org/tracker/joomla-cms/37910</a>.</sub>

richard67 commented:
@nikosdion Hmm, drone reports a javascript code style error here https://ci.joomla.org/joomla/joomla-cms/55438/1/26 :
```
/********/src/build/media_source/plg_system_webauthn/js/management.es6.js
  376:5  error  Arrow function expected no return value  consistent-return
```
Could you fix that?

Commit:
Make the bots happy again

nikosdion commented:
@richard67 There you go

richard67 commented:
I've restored the previous human test result because the change after that was not really functional but code style.

laoneo commented on plugins/system/webauthn/src/Authentication.php:
```diff
@@ -0,0 +1,586 @@
+	{
+		try
+		{
+			$this->app = $app ?? Factory::getApplication();
```
This is not needed anymore as the service provider can set the app now with setApplication and it is accessible with getApplication.

nikosdion replied:
This is not the plugin class.

laoneo commented on plugins/system/webauthn/src/Extension/Webauthn.php:
```diff
@@ -0,0 +1,199 @@
+ *
+ * @since  4.0.0
+ */
+class Webauthn extends CMSPlugin implements SubscriberInterface
```
Can this be marked as final

nikosdion replied:
Agreed and committed.

laoneo commented on plugins/system/webauthn/src/Extension/Webauthn.php:
```diff
@@ -0,0 +1,199 @@
+	 * @var   CMSApplication
+	 * @since __DEPLOY_VERSION__
+	 */
+	protected $app;
```
Better to use now setApplication

nikosdion replied:
Agreed and committed.

laoneo committed:
Merge branch '4.2-dev' into feature/webauthn-refactor-2

Commit:
Merge branch '4.2-dev' into feature/webauthn-refactor-2

Commit:
Code polishing

* Marking classes final
* Use setApplication / getApplication in the plugin class
* Remove unused `$db` from the plugin class

Commit:
Merge remote-tracking branch 'nikosdion/feature/webauthn-refactor-2' into feature/webauthn-refactor-2

nikosdion commented:
Code polishing done per @laoneo suggestions.

brianteeman commented:
I have tested this item :white_check_mark: successfully on c03b191350d4003b9ca3ba05a0c4c423c6d52a35<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37910">issues.joomla.org/tracker/joomla-cms/37910</a>.</sub>

richard67 commented:
@Kostelano Could you test this PR again? I know, you did already a few times, but there have been some changes which were not code style only. Thanks in advance.

richard67 commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37910">issues.joomla.org/tracker/joomla-cms/37910</a>.</sub>

nikosdion commented:
/me doing a happy dance!

Guys, this is awesome. Between this and MFA we can market Joomla 4.2 as the security and privacy enhancement update. Yay! Joomla rocks!

roland-d committed:
Merge branch '4.2-dev' into feature/webauthn-refactor-2

HLeithner reviewed (changes requested):
This PR fails for my test. 

```
Call stack
--
# | Function | Location
1 | () | JROOT/plugins/system/webauthn/src/PluginTraits/UserProfileFields.php:233
2 | Joomla\Plugin\System\Webauthn\Extension\Webauthn->canEditUser() | JROOT/plugins/system/webauthn/src/PluginTraits/UserProfileFields.php:132
3 | Joomla\Plugin\System\Webauthn\Extension\Webauthn->onContentPrepareForm() | JROOT/libraries/vendor/joomla/event/src/Dispatcher.php:486
4 | Joomla\Event\Dispatcher->dispatch() | JROOT/libraries/src/Application/EventAware.php:115
5 | Joomla\CMS\Application\WebApplication->triggerEvent() | JROOT/libraries/src/MVC/Model/FormBehaviorTrait.php:179
6 | Joomla\CMS\MVC\Model\FormModel->preprocessForm() | JROOT/administrator/components/com_users/src/Model/UserModel.php:216
7 | Joomla\Component\Users\Administrator\Model\UserModel->preprocessForm() | JROOT/libraries/src/MVC/Model/FormBehaviorTrait.php:117
8 | Joomla\CMS\MVC\Model\FormModel->loadForm() | JROOT/administrator/components/com_users/src/Model/UserModel.php:126
9 | Joomla\Component\Users\Administrator\Model\UserModel->getForm() | JROOT/libraries/src/MVC/View/AbstractView.php:153
10 | Joomla\CMS\MVC\View\AbstractView->get() | JROOT/administrator/components/com_users/src/View/User/HtmlView.php:95
11 | Joomla\Component\Users\Administrator\View\User\HtmlView->display() | JROOT/libraries/src/MVC/Controller/BaseController.php:708
12 | Joomla\CMS\MVC\Controller\BaseController->display() | JROOT/administrator/components/com_users/src/Controller/DisplayController.php:142
13 | Joomla\Component\Users\Administrator\Controller\DisplayController->display() | JROOT/libraries/src/MVC/Controller/BaseController.php:746
14 | Joomla\CMS\MVC\Controller\BaseController->execute() | JROOT/libraries/src/Dispatcher/ComponentDispatcher.php:146
15 | Joomla\CMS\Dispatcher\ComponentDispatcher->dispatch() | JROOT/libraries/src/Component/ComponentHelper.php:389
16 | Joomla\CMS\Component\ComponentHelper::renderComponent() | JROOT/libraries/src/Application/AdministratorApplication.php:144
17 | Joomla\CMS\Application\AdministratorApplication->dispatch() | JROOT/libraries/src/Application/AdministratorApplication.php:187
18 | Joomla\CMS\Application\AdministratorApplication->doExecute() | JROOT/libraries/src/Application/CMSApplication.php:296
19 | Joomla\CMS\Application\CMSApplication->execute() | JROOT/administrator/includes/app.php:63
20 | require_once() | JROOT/administrator/index.php:32
```

The reason for this is the conflict between this PR and the already merged PR https://github.com/joomla/joomla-cms/pull/38060 which implements a setApplication for plugins using a service Provider.

I'm not 100% sure where the conflict exists except that it seems that $application hasn't been injected (not sure why this worked before, at least I would expect it worked before the last 4.2-dev merge).

@nikosdion I would expect you find the problem faster then me.

For the moment RTC have to be removed @richard67 till we have an application again, and maybe we have to check if other plugin PRs are effected too.

Commit:
Blind fix

Currently #38060 has broken everything it seems?

nikosdion commented:
Well, https://github.com/joomla/joomla-cms/pull/38060 has broken plenty of things as it looks like.

The MFA is currently broken as well. As a result I cannot test my blind fix which I just committed.

brianteeman commented:
That pr breaks quite a bit.

HLeithner commented:
> Well, #38060 has broken plenty of things as it looks like.
> 
> The MFA is currently broken as well. As a result I cannot test my blind fix which I just committed.

based on the other pr you shouldn't extend the constructor instead you should implement the ApplicationAwareInterface and inject the Application in the service provider.

nikosdion commented:
Disabling the MFA plugins by editing the database I was able to confirm that my fix worked.

The real problem is that #38060 should have never been merged and I am looking at you @roland-d . Architectural work MUST ALWAYS be split in separate PRs: one PR to do the architectural work with 100% backwards compatibility and adding deprecations for what is being replaced. Then a number of PRs making changes in the affected areas with thorough testing.

For example, I see that in plg_multifactor_auth @laoneo just removed $this->app and added the application by injection. But he DID NOT EVEN CHECK that $this->app is no longer present in the plugins/multifactorauth/webauthn tree! Had he done this basic check he'd have seen that there's tmpl/default.php which is still using it. This is what was causing a problem for me trying to fix the problem @HLeithner reported.

Whenever I see something **AS BASIC AS THAT** not being done in a PR I only have one recommendation to make: revert the PR as soon as possible. It was not ready to get merged.

When I submit a PR you nitpick on it and I have to spend 10x the time I spent writing code just to address your pointless nitpicking. Yet when one of your “inner circle” submits a PR which breaks userland what do you do? Oh, right, zero discussion, just merge it and break userland! Then you get upset when I call you hypocrites and I am labelled "difficult to work with". To be clear: what I want is that **EVERYONE's Pull Requests** undergo the SAME amount of scrutiny and that amount of scrutiny has to be close to or equal to the scrutiny you put mine. Otherwise you ARE hypocrites and you are not doing right by the community who trusted you with commit rights.

Now you are telling me off for extending the constructor. You know what? Fuck it.

HLeithner commented:
> Now you are telling me off for extending the constructor. You know what? Fuck it.

I was not involved in the other pr, I only looked how he did it and suggested it. Aside this yes the PR is really ugly and wrong. Reverting it doesn't sound like a bad idea problem at it is that beta2 has released and we should fix it in a way that we keep the interface and the trait. The rest should be done as you said.

joomdonation commented:
> I did try and flag that it was way too late in the release cycle for the type of changes in that pr [#38103 (comment)](https://github.com/joomla/joomla-cms/pull/38103#issuecomment-1161382860)

Actually, the PR https://github.com/joomla/joomla-cms/pull/38103 is fine as it is. It is kind of bug fix for the ugly code, it does not break anything and could be merged, even for patch release. But It is up to maintainer to decide.

brianteeman commented:
It was a general flag and not specific. I dont have the skills to comment on the specific code but I do have 15 years of watching that type of pr being merged without tests just before releases and watching them blow up.

richard67 commented:
@nikosdion Don't be childish. Please re-open this PR.

nikosdion commented:
Dude, this is now how you de-escalate :)

When my ISP decides to fix the problem which causes no service to the entire area I will re-open the PR. Can’t do it on mobile as I can’t push the branch again. 

On another note, once I do that please do take over any necessary changes. I hate having to do that to you but it’s for medical reasons. 

I believe in full transparency and I have no problem explaining this further. I’m on ADHD medication which helps me deal with memory issues and impulsiveness. It does, however, affect the autonomic nervous system. As a result I got a dangerously high resting heart rate the last two weeks. We tried lowering my dose, that only made me irritable (well, as you know) so now I have to go through a whole lot of tests to see if I can also get beta blockers. Until then I am to reduce my stress levels which unfortunately means I can’t contribute to Joomla, among other things I like but stress me out that I have to forego temporarily.

nikosdion commented:
@richard67 There you go. Good luck, folks.

richard67 commented:
Thanks, and all the best for you.

laoneo commented:
@nikosdion thanks for sharing your private situation. Now I can put your words in the right context. If you want we can have a private talk and I can explain you the current situation why all this architecture work is happening. And I hope you understand that it was never my intention to add BC breaks. But at the end of the day we are all humans and do make mistakes, important is that we solve them. And final words from me, don't blame or shitstorm Roland, I was the person how made that stuff and put pressure on him for merges.

HLeithner committed:
Bring application injection in sync with core

richard67 commented:
> @richard67 I created the [nikosdion#9](https://github.com/nikosdion/joomla-cms/pull/9) pr which (also blindly) should bring this PR in sync with the request of the cms plugins.

@HLeithner I can't merge in @nikosdion repo, so either he merges it, or I have to take over this PR here by making a new one (using a branch in my repo based on Nik's branch here so commit history would not get lost).

Commit:
Merge pull request #9 from HLeithner/hello-windows

Bring application injection in sync with core

nikosdion commented:
I can merge PR in my repo. Just at-me so I know when I have to.

wilsonge committed:
Remove whitespace

richard67 commented:
@wilsonge As you can commit directly, could you commit also a use statement for the Factory to plugins/system/webauthn/services/provider.php? Currently the api tests fail at the login after the installation with a "Class 'Factory' not found" error.

wilsonge committed:
Add use statement

richard67 commented:
@HLeithner Is there something more to do after your PR for this branch has been merged by Nik? If not and this PR is ok now: Could you dismiss or resolve your review so that GitHub doesn't show anymore "Changes requested" in the list of PRs? It's something GitHub does when there are unresolved reviews, not to be mixed up with the "Updates Requested" label which we set.

HLeithner commented:
I only die the low hanging stuff but didn't hat the time to check why the redirection to the MFA didn't work when fido is used

richard67 commented:
> That didn't fix the 2fa problem yet. For this we have to find out why it doesn't get redirect to the correct page

@HLeithner Ah, I just see there is still something to be done. Or is that the thing solved with PR #38132 ? I don't think it is, but I'm asking to be sure.

HLeithner commented:
@brianteeman would you be so kind and test if the error exists for you too? maybe it's my installation

brianteeman commented:
Wont get a chance until tomorrow am

Kostelano commented:
I updated Joomla with this PR. Deleted / Created a fingerprint key, everything works. I couldn't get any errors.

HLeithner commented:
> I updated Joomla with this PR. Deleted / Created a fingerprint key, everything works. I couldn't get any errors.

not only fingerprint, please add a 2fa authentication like google authenticator

brianteeman commented:
testing now

brianteeman commented:
Completely new install directly from nikosdion:feature/webauthn-refactor-2

logged into the admin and went to my user account and the W3C Web Authentication (WebAuthn) Login tab

successfully added a windows hello pin ✅

went to the Multi-factor Authentication tab and clicked on the Add a new web authentication

error has occurred ❌

<html>
<body>
<!--StartFragment--><h1 style="box-sizing: border-box; margin-top: 0px; margin-bottom: 0.5rem; font-weight: 400; line-height: 1.2; color: var(--template-bg-dark); font-size: 1.65rem; font-family: Roboto, -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, &quot;Helvetica Neue&quot;, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(237, 242, 247); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">An error has occurred.</h1><blockquote class="blockquote" style="box-sizing: border-box; margin: 0px 0px 1rem; font-size: 1.25rem; color: rgb(73, 80, 87); font-family: Roboto, -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, &quot;Helvetica Neue&quot;, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(237, 242, 247); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><span class="badge bg-secondary" style="box-sizing: border-box; display: inline-block; padding: 0.3rem 0.2rem; font-size: 0.75rem; font-weight: 700; line-height: 1; color: rgb(255, 255, 255); text-align: center; white-space: nowrap; vertical-align: baseline; border-radius: 0.2rem; --bg-opacity: 1; background-color: rgba(var(--secondary-rgb), var(--bg-opacity))  !important; margin-bottom: 0px;">0</span><span> </span>Call to a member function getDocument() on null</blockquote><div style="box-sizing: border-box; color: rgb(73, 80, 87); font-family: Roboto, -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, &quot;Helvetica Neue&quot;, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(237, 242, 247); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">

Call stack
--
# | Function | Location
1 | () | JROOT/plugins/multifactorauth/webauthn/tmpl/default.php:36
2 | include() | JROOT/plugins/multifactorauth/webauthn/src/Extension/Webauthn.php:146
3 | Joomla\Plugin\Multifactorauth\Webauthn\Extension\Webauthn->onUserMultifactorGetSetup() | JROOT/libraries/vendor/joomla/event/src/Dispatcher.php:486
4 | Joomla\Event\Dispatcher->dispatch() | JROOT/administrator/components/com_users/src/Model/MethodModel.php:102
5 | Joomla\Component\Users\Administrator\Model\MethodModel->getRenderOptions() | JROOT/administrator/components/com_users/src/View/Method/HtmlView.php:118
6 | Joomla\Component\Users\Administrator\View\Method\HtmlView->display() | JROOT/administrator/components/com_users/src/Controller/MethodController.php:120
7 | Joomla\Component\Users\Administrator\Controller\MethodController->add() | JROOT/libraries/src/MVC/Controller/BaseController.php:746
8 | Joomla\CMS\MVC\Controller\BaseController->execute() | JROOT/administrator/components/com_users/src/Controller/MethodController.php:75
9 | Joomla\Component\Users\Administrator\Controller\MethodController->execute() | JROOT/libraries/src/Dispatcher/ComponentDispatcher.php:146
10 | Joomla\CMS\Dispatcher\ComponentDispatcher->dispatch() | JROOT/libraries/src/Component/ComponentHelper.php:389
11 | Joomla\CMS\Component\ComponentHelper::renderComponent() | JROOT/libraries/src/Application/AdministratorApplication.php:144
12 | Joomla\CMS\Application\AdministratorApplication->dispatch() | JROOT/libraries/src/Application/AdministratorApplication.php:187
13 | Joomla\CMS\Application\AdministratorApplication->doExecute() | JROOT/libraries/src/Application/CMSApplication.php:296
14 | Joomla\CMS\Application\CMSApplication->execute() | JROOT/administrator/includes/app.php:63
15 | require_once() | JROOT/administrator/index.php:32

</div><!--EndFragment-->
</body>
</html>

HLeithner commented:
Oh that's a new issue^^

richard67 commented:
Looks like the issue fixed by PR #38132 .

nikosdion commented:
My recommendation is to merge this PR as it was in commit https://github.com/joomla/joomla-cms/pull/37910/commits/fbfba3cdae421a3de349081d2fb1db555ec27ed0 which was fully tested and working. You can then try to move it from $this->app to whatever half-baked idea you pushed into 4.2 without any discussion.

richard67 committed:
Merge branch '4.2-dev' into feature/webauthn-refactor-2

richard67 commented:
@brianteeman Could you test again? I think your issue has been fixed with PR #38132 , and that just has been merged and now I have update the branch of this PR here so it has that change, too.

brianteeman commented:
@richard67 that resolved that issue. Now what was the issue @HLeithner has - i cant see it here

richard67 commented:
> @richard67 that resolved that issue. Now what was the issue @HLeithner has - i cant see it here

That's a good step forward.

HLeithner commented:
I also installed the current version and I have the same result. Additional I got an error message when trying to save the user with 2 x webauthn + 1x 2fa methods which seems to be related to the privacy action log:
`Save failed with the following error: Field 'extensions' doesn't have a default value`
After disabling the user action log plugin I was able to save the user again.... maybe unrelated? didn't checked the code yet if it creates an action log entry.

HLeithner commented on plugins/system/webauthn/src/PluginTraits/AjaxHandlerLogin.php:
```diff
@@ -118,153 +133,195 @@ public function onAjaxWebauthnLogin(): void
+
+		// Run the user plugins. They CAN block login by returning boolean false and setting $response->error_message.
+		PluginHelper::importPlugin('user');
+		$event   = AbstractEvent::create('onUserLogin', [(array) $response, $options]);
```
```suggestion
		$event   = new \Joomla\Event\Event('onUserLogin', [(array) $response, $options]);
```
No idea how this ever worked, using the AbstractEvent creator ends into an error log entry 'Received login failure. Message: No subject given for the onUserAfterLogin event'.

The `AbstractEvent::create` function needs a subject, it doesn't need any if you use the deprecated `tiggerEvent` function or directly create the event...

nikosdion replied:
This was a wrong search & replace. I'm fixing it now.

HLeithner commented:
The "the user action log plugin" problem is a php 8.1 problem.

nikosdion commented:
Give me half an hour or so to check all events creations. I think I have an idea about what happened.

HLeithner commented:
> Give me half an hour or so to check all events creations. I think I have an idea about what happened.

sounds good.

Commit:
Fix wrong event creation in AjaxHandlerLogin

nikosdion commented:
@HLeithner Done.

When moving the plugin to concrete events I made a mistake in the way I was creating the events. I have updated the code to work similar to how I had updated triggerEvent to work so when we create concrete events for user login / login failure nothing will break.

Commit:
Merge branch '4.2-dev' into feature/webauthn-refactor-2

HLeithner commented on plugins/system/webauthn/src/Hotfix/AndroidKeyAttestationStatementSupport.php:
```diff
@@ -0,0 +1,270 @@
+<?php
+/**
+ * @package         Joomla.Plugin
+ * @subpackage      System.Webauthn
+ *
+ * @copyright   (C) 2022 Open Source Matters, Inc. <https://www.joomla.org>
+ * @license         GNU General Public License version 2 or later; see LICENSE.txt
+ */
```
Upstream is MIT right? we can't simply re-license this

HLeithner commented on plugins/system/webauthn/src/Hotfix/FidoU2FAttestationStatementSupport.php:
```diff
@@ -0,0 +1,230 @@
+<?php
+/**
+ * @package         Joomla.Plugin
+ * @subpackage      System.Webauthn
+ *
+ * @copyright   (C) 2022 Open Source Matters, Inc. <https://www.joomla.org>
+ * @license         GNU General Public License version 2 or later; see LICENSE.txt
+ */
```
License issue like the other files in this directory

HLeithner commented:
@nikosdion can you change the 3 files having a license issue because you copy-paste it from upstream, they have to stay MIT with the correct reference to at least that's my understanding.

HLeithner commented:
Beside the license issue, I tested this successful.

nikosdion commented:
When you are combining GPL code with any GPL-compatible (e.g. MIT licensed) code the entire combined product becomes GPL. So even if we note these files as MIT licensed since they are part of our GPLv2 product they are effectively licensed under GPL when the software is distributed to the users. See https://www.gnu.org/licenses/gpl-faq.html#WhatDoesCompatMean

Sure, I can change the license of these files but it's pointless to the effect that the mere act of putting them on GitHub as part of Joomla constitutes distribution therefore they are effectively licensed as GPLv2 (even though they were forked off MIT code).

Commit:
License change

HLeithner commented:
> License notes changed per your request

Thanks, better give proper contribution then nothing even if it's maybe legal but I'm not a layer so I hope that fits our needs.

HLeithner committed:
Merge branch '4.2-dev' into feature/webauthn-refactor-2

HLeithner commented:
So I merge this for now and if we find problems we can fix them after the psr-12 merge.

thanks

brianteeman commented:
Sorry to be a pain but this doesnt satisfy the terms of the licence

> The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

brianteeman commented:
> When you are combining GPL code with any GPL-compatible (e.g. MIT licensed) code the entire combined product becomes GPL. So even if we note these files as MIT licensed since they are part of our GPLv2 product they are effectively licensed under GPL when the software is distributed to the users. See https://www.gnu.org/licenses/gpl-faq.html#WhatDoesCompatMean
> 
> Sure, I can change the license of these files but it's pointless to the effect that the mere act of putting them on GitHub as part of Joomla constitutes distribution therefore they are effectively licensed as GPLv2 (even though they were forked off MIT code).

The combined product is GPLv2 but the original copyright on the individual files does not change and the terms of that licence must still be adhered to within those files.

HLeithner commented:
Can you create a pr to fix this? I think you have the best knowledge on this topic?
